core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2426,20 +2425,22 @@ object Stream {
      * For example, `merge` is implemented by calling `unconsAsync` on each stream, racing the
      * resultant `AsyncPull`s, emitting winner of the race, and then repeating.
      */
-    def unconsAsync(implicit ec: ExecutionContext, F: Effect[F]): Pull[F,Nothing, AsyncPull[F,Option[(Segment[O,Unit], Stream[F,O])]]] = {
+    def unconsAsync(implicit ec: ExecutionContext, F: Effect[F]): Pull[F,Nothing,AsyncPull[F,Option[(Segment[O,Unit], Stream[F,O])]]] = {
       type UO = Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]
-
-      Pull.fromFreeC {
-         val p = async.Promise.unsafeCreate[F, Either[Throwable, Option[(Segment[O,Unit], Stream[F,O])]]]
-        Algebra.getScope[F, Nothing] flatMap { scope =>
-          val runStep =
-            Algebra.compileScope(
-              scope
-              , Algebra.uncons(self.get).flatMap(Algebra.output1(_))
-              , None : UO
-            ){ (_, uo) => uo.asInstanceOf[UO] } map { _ map { case (hd, tl) => (hd, fromFreeC(tl)) }}
-
-          Algebra.eval(async.fork(F.flatMap(F.attempt(runStep))(x => async.fork(p.complete(x))))) map { _ => AsyncPull.readAttemptPromise(p) }
+      type Res = Option[(Segment[O,Unit], Stream[F,O])]
+
+      Pull.eval(Promise.empty[F, Either[Throwable, Res]]) flatMap { p =>
+        Pull.fromFreeC {
```

Comment:
You could inline this `Pull`:
```scala
Pull.fromFreeC { Algebra.eval[F, Nothing, Promise[F, Res]](async.Promise.empty).flatMap { p =>
```

Reply:
yeah that's a good idea, I'll do that

Reply:
In fact, the whole operation seems quite similar to `start`, so maybe it's worth thinking for a while how to make it a bit nicer